apps/calendar/js/calendar_observer.js

```diff
@@ -0,0 +1,63 @@
+define(function(require, exports, module) {
+'use strict';
+
+var EventEmitter2 = require('ext/eventemitter2');
+var debug = require('debug')('calendar_observer');
+var forEach = require('object').forEach;
+var nextTick = require('next_tick');
+
+var observer = module.exports = new EventEmitter2();
+
+// TODO: convert these dependencies into static modules to avoid ugly injection
+// injected later to avoid circular dependencies
+observer.calendarStore = null;
+
+observer.calendarList = null;
+
+observer.init = function() {
+  this.calendarList = {};
+  this._cacheCalendar = this._cacheCalendar.bind(this);
+  this.calendarStore.on('add', this._cacheCalendar);
+  this.calendarStore.on('update', this._cacheCalendar);
+  this._purgeCalendar = this._purgeCalendar.bind(this);
+  this.calendarStore.on('remove', this._purgeCalendar);
+  this.calendarStore.all().then(calendars => {
+    forEach(calendars, this._cacheCalendar);
+  });
+};
+
+observer._cacheCalendar = function(id, calendar) {
+  debug('Will cache new calendar', calendar);
+  return this.calendarStore.providerFor(calendar).then(provider => {
+    var capabilities = provider.calendarCapabilities(calendar);
+    this.calendarList[id] = {
+      calendar: calendar,
+      capabilities: capabilities
+    };
```

Comment:
since the `providerFor` is async there is a small chance for a race condition here if `remove` is called before it updates the `calendarList`